In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sns

sns.set_theme()

# jupyter notebook full-width display
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:100% !important; }</style>"))

# no text wrapping
display(HTML("<style>.dataframe td { white-space: nowrap; }</style>"))

# pandas formatting
pd.set_option('display.float_format', '{:.3f}'.format)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 200)
pd.set_option('display.max_colwidth', 400)

In [2]:
df_SD = pd.read_pickle('df_SD.pickle')
df_FD = pd.read_pickle('df_FD.pickle')
df_LF = pd.read_pickle('df_LF.pickle')
df_LF_grouped = pd.read_pickle('df_LF_grouped.pickle')
df_Site = pd.read_pickle('df_Site.pickle')
df_TrapSupervisors = pd.read_pickle('df_TrapSupervisors.pickle')

# checking stuff

### why don't these match?
* answer: they do, but there is no matching sample to link them

In [3]:
df_LF[(df_LF.yy==2019) & (df_LF.mm==5) & (df_LF.dd==28) & (df_LF.site=="47")]

yy  mm  dd  Time     river  week site    loc period  wt_lbs  wt_kg  \
11195  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11196  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11197  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11198  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11199  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11200  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11201  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11202  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11203  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11204  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   
11205  2019   5  28   NaN  MARGAREE     5   47  UPPER     PM  80.900 36.696   

       lgth  freq  Flbin   DATETIME SITE1 SITE2 SITE3          id  \
11195   215     1    215 2019-05-28    47  <NA>  <NA>  2019052847   
11196   220     6    220 2019-05-28    47  <NA>  <NA>  2019052847   
11197   225    16    225 2019-05-28    47  <NA>  <NA>  2019052847   
11198   230    26    230 2019-05-28    47  <NA>  <NA>  2019052847   
11199   235    39    235 2019-05-28    47  <NA>  <NA>  2019052847   
11200   240    51    240 2019-05-28    47  <NA>  <NA>  2019052847   
11201   245    35    245 2019-05-28    47  <NA>  <NA>  2019052847   
11202   250    27    250 2019-05-28    47  <NA>  <NA>  2019052847   
11203   255    17    255 2019-05-28    47  <NA>  <NA>  2019052847   
11204   260     5    260 2019-05-28    47  <NA>  <NA>  2019052847   
11205   265     1    265 2019-05-28    47  <NA>  <NA>  2019052847   

       length_bin_id FLAG_SITE FLAG_AM_PM_PERIOD  \
11195         21.500       NaN               NaN   
11196         22.000       NaN               NaN   
11197         22.500       NaN               NaN   
11198         23.000       NaN               NaN   
11199         23.500       NaN               NaN   
11200         24.000       NaN               NaN   
11201         24.500       NaN               NaN   
11202         25.000       NaN               NaN   
11203         25.500       NaN               NaN   
11204         26.000       NaN               NaN   
11205         26.500       NaN               NaN   

      FLAG_AM_PM_PERIOD_DISCREPANCIES FLAG_NO_MATCHING_SAMPLE  
11195                             NaN                    True  
11196                             NaN                    True  
11197                             NaN                    True  
11198                             NaN                    True  
11199                             NaN                    True  
11200                             NaN                    True  
11201                             NaN                    True  
11202                             NaN                    True  
11203                             NaN                    True  
11204                             NaN                    True  
11205                             NaN                    True

In [4]:
df_SD[(df_SD.DATETIME.dt.year==2019) & (df_SD.DATETIME.dt.month==5) & (df_SD.DATETIME.dt.day==28)]

DIST   RIVER                 NAME code  GEAR SITE_NO no_nets  YEAR  MM  \
14959     2  SWMARG     Martin E Cameron  NaN    81      12       1  2019   5   
14960     2  SWMARG        Robert Peters  NaN    81      25       1  2019   5   
14961     2  SWMARG      Pierre Chiasson  NaN    81      26       1  2019   5   
14962     2  SWMARG    Gerard MacFarlane  NaN    81      28       1  2019   5   
14963     2  SWMARG       Daniel Stewart  NaN    81      33       1  2019   5   
14964     2  SWMARG       Stewart Gillis  NaN    81      41       1  2019   5   
14965     2  SWMARG       Finley Stewart  NaN    81      44     NaN  2019   5   
14966     2  SWMARG        Mary E Gillis  NaN    81      49       1  2019   5   
14967     2  SWMARG      Bruce MacLellan  NaN    81      51       1  2019   5   
14968     2  SWMARG  Elizabeth MacKinnon  NaN    81      62       1  2019   5   
14969     2  SWMARG      Kevin MacKinnon  NaN    81      64       1  2019   5   

       DD  Week  catch_lbs  catch_kg hours_fished   zone   last_name comments  \
14959  28     5    350.000   158.800            5  lower     Cameron      NaN   
14960  28     5     25.000    11.300            8  lower      Peters      NaN   
14961  28     5    200.000    90.700            7  lower    Chiasson      NaN   
14962  28     5    200.000    90.700            8  upper  MacFarlane      NaN   
14963  28     5    150.000    68.000            8  upper     Stewart      NaN   
14964  28     5     30.000    13.600            7  upper      Gillis      NaN   
14965  28     5    300.000   136.100           14  upper     Stewart      NaN   
14966  28     5    700.000   317.500           14  upper      Gillis      NaN   
14967  28     5     15.000     6.800            6  upper   MacLellan      NaN   
14968  28     5    500.000   226.800           10  upper   MacKinnon      NaN   
14969  28     5    120.000    54.400            6  upper   Mackinnon      NaN   

       bycatch_sbass  bycatch_shad bycatch_other   DATETIME  SITE1 SITE2  \
14959            NaN           NaN           NaN 2019-05-28     12  <NA>   
14960            NaN           NaN           NaN 2019-05-28     25  <NA>   
14961            NaN           NaN           NaN 2019-05-28     26  <NA>   
14962            NaN           NaN       1 perch 2019-05-28     28  <NA>   
14963            NaN           NaN           NaN 2019-05-28     33  <NA>   
14964            NaN           NaN           NaN 2019-05-28     41  <NA>   
14965            NaN           NaN           NaN 2019-05-28     44  <NA>   
14966            NaN           NaN           NaN 2019-05-28     49  <NA>   
14967            NaN           NaN           NaN 2019-05-28     51  <NA>   
14968            NaN           NaN           NaN 2019-05-28     62  <NA>   
14969            NaN           NaN           NaN 2019-05-28     64  <NA>   

                      remarks          id  total_fish_preserved  \
14959                     NaN  2019052812                31.000   
14960                     NaN  2019052825                   NaN   
14961                     NaN  2019052826                   NaN   
14962  bycatch_other: 1 perch  2019052828                   NaN   
14963                     NaN  2019052833                   NaN   
14964                     NaN  2019052841                   NaN   
14965                     NaN  2019052844                   NaN   
14966                     NaN  2019052849                   NaN   
14967                     NaN  2019052851                   NaN   
14968                     NaN  2019052862                36.000   
14969                     NaN  2019052864                   NaN   

       total_fish_measured AM_PM_PERIOD  wt_lbs FLAG_DATETIME  \
14959              195.000           AM  71.900           NaN   
14960                  NaN          NaN     NaN           NaN   
14961                  NaN          NaN     NaN           NaN   
14962                  NaN          NaN     NaN           NaN   
14963             

### kevin followup question: how many ambiguous matches don't have a matching sample because there are multiple sample matches

In [5]:
ambiguous = set()
for i in list(df_SD[df_SD.id>2024000000].id):
    while i > 2024000000:
        i -= 1000000000
    ambiguous.add(i)
    
ambiguous = list(ambiguous)

In [6]:
# EXACT MATCHES = NONE

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # no exact matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY) & (df_LF.site==str(SITE))])
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY) & (df_FD.SITE==SITE)])


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

In [7]:
# POTENTIAL MATCHES WITH MESSY SITE NAMES = NONE

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # partial matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY) & (df_LF.site.str.contains(str(SITE))==True)])
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY) & (df_FD.SITE.str.contains(str(SITE))==True)])


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

In [8]:
# A FEW MATCHES WHEN SITE IS NOT CONSIDERED
# maybe there the site number was written down wrong, some partial matches (excluding site)

for ambiguous_sample in ambiguous:
    YEAR, MONTH, DAY, SITE = (
        ambiguous_sample//1000000, 
        ambiguous_sample//10000 - 100*(ambiguous_sample//1000000), 
        ambiguous_sample//100 - 100*(ambiguous_sample//10000), 
        ambiguous_sample - 100*(ambiguous_sample//100)
    )
    print()
    print(YEAR, MONTH, DAY, SITE)
    # very partial matches
    display(df_LF[(df_LF.yy==YEAR) & (df_LF.mm==MONTH) & (df_LF.dd==DAY)].head())
    display(df_FD[(df_FD.YEAR==YEAR) & (df_FD.MM==MONTH) & (df_FD.DD==DAY)].head())


2004 6 10 1


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []


1997 6 12 58


yy  mm  dd  Time        river  week site    loc period  wt_lbs  wt_kg  \
3024  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3025  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3026  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3027  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   
3028  1997   6  12   NaN  SW MARGAREE     7   26  LOWER     AM 118.000 53.500   

      lgth  freq  Flbin   DATETIME SITE1 SITE2 SITE3          id  \
3024   220     3    220 1997-06-12    26  <NA>  <NA>  1997061226   
3025   225     4    225 1997-06-12    26  <NA>  <NA>  1997061226   
3026   230    19    230 1997-06-12    26  <NA>  <NA>  1997061226   
3027   235    52    235 1997-06-12    26  <NA>  <NA>  1997061226   
3028   240    70    240 1997-06-12    26  <NA>  <NA>  1997061226   

      length_bin_id FLAG_SITE FLAG_AM_PM_PERIOD  \
3024         22.000       NaN               NaN   
3025         22.500       NaN               NaN   
3026         23.000       NaN               NaN   
3027         23.500       NaN               NaN   
3028         24.000       NaN               NaN   

     FLAG_AM_PM_PERIOD_DISCREPANCIES FLAG_NO_MATCHING_SAMPLE  
3024                             NaN                     NaN  
3025                             NaN                     NaN  
3026                             NaN                     NaN  
3027                             NaN                     NaN  
3028                             NaN                     NaN

YEAR  MM  DD  WEEK SITE PERIOD CONDITION  FISH_NO  FL_WET  FL_FROZEN  \
15016  1997   6  12     7   26     AM    Frozen        1    <NA>        230   
15017  1997   6  12     7   26     AM    Frozen        2    <NA>        224   
15018  1997   6  12     7   26     AM    Frozen        3    <NA>        272   
15019  1997   6  12     7   26     AM    Frozen        4    <NA>        212   
15020  1997   6  12     7   26     AM    Frozen        5    <NA>        220   

       FL_STD  WEIGHT SPECIES SEX  MATURITY  GONAD_WEIGHT Ager_1  AGE_1  \
15016 238.000 171.000       A   M         4           NaN    NaN      3   
15017 232.000 166.000       A   M         4           NaN    NaN      3   
15018 280.000 307.000       B   F         4        39.900    NaN      7   
15019 220.000 138.000       A   M         4           NaN    NaN      3   
15020 228.000 143.000       A   M         4           NaN    NaN      3   

       FSP_1 Comments_1 Ager_2  AGE_2  FSP_2 Comments_2 Ager_3  AGE_3  FSP_3  \
15016      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15017      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15018      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15019      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
15020      3        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   

      Comments_3 Envelop.Comments   DATETIME AGE_notes_1 FSP_notes_1  \
15016        NaN              NaN 1997-06-12           3           3   
15017        NaN              NaN 1997-06-12           3           3   
15018        NaN              NaN 1997-06-12           7           3   
15019        NaN              NaN 1997-06-12           3           3   
15020        NaN              NaN 1997-06-12           3           3   

      AGE_notes_2 FSP_notes_2 AGE_notes_3 FSP_notes_3 SITE_notes  SITE1 SITE2  \
15016         NaN         NaN         NaN         NaN         26     26  <NA>   
15017         NaN         NaN         NaN         NaN         26     26  <NA>   
15018         NaN         NaN         NaN         NaN         26     26  <NA>   
15019         NaN         NaN         NaN         NaN         26     26  <NA>   
15020         NaN         NaN         NaN         NaN         26     26  <NA>   

      SITE3  fish_length                                         remarks  \
15016  <NA>          230  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15017  <NA>          224  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15018  <NA>          272  SITE_notes: 26; AGE_notes_1: 7; FSP_notes_1: 3   
15019  <NA>          212  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   
15020  <NA>          220  SITE_notes: 26; AGE_notes_1: 3; FSP_notes_1: 3   

               id FLAG_SITE FLAG_AM_PM_PERIOD FLAG_SEX FLAG_MATURITY  \
15016  1997061226       NaN               NaN      NaN           NaN   
15017  1997061226       NaN               NaN      NaN           NaN   
15018  1997061226       NaN               NaN      NaN           NaN   
15019  1997061226       NaN               NaN      NaN           NaN   
15020  1997061226       NaN               NaN      NaN           NaN   

      FLAG_FSP_1 FLAG_FL_STD FLAG_FL_WET_FROZEN FLAG_WEIGHT_OUTLIER  \
15016        NaN         NaN                NaN                 NaN   
15017        NaN         NaN                NaN                 NaN   
15018        NaN         NaN                NaN                 NaN   
15019        NaN         NaN                NaN                 NaN   
15020        NaN         NaN                NaN                 NaN   

      FLAG_GONAD_OUTLIER FLAG_MULTIPLE_SAMPLE_POSSIBILITIES  \
15016                NaN                                NaN   
15017                NaN                                NaN   
15018                NaN                                NaN   
15019                NaN                                NaN   
15020                NaN                        


1988 5 23 48


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

YEAR  MM  DD  WEEK SITE PERIOD CONDITION  FISH_NO  FL_WET  FL_FROZEN  \
4503  1988   5  23  <NA>   12    NaN     Fresh        1     253       <NA>   
4504  1988   5  23  <NA>   12    NaN     Fresh        2     258       <NA>   
4505  1988   5  23  <NA>   12    NaN     Fresh        3     245       <NA>   
4506  1988   5  23  <NA>   12    NaN     Fresh        4     250       <NA>   
4507  1988   5  23  <NA>   12    NaN     Fresh        5     245       <NA>   

      FL_STD  WEIGHT SPECIES SEX  MATURITY  GONAD_WEIGHT Ager_1  AGE_1  FSP_1  \
4503 253.000 208.000       A   M         4           NaN    NaN      4      4   
4504 258.000 236.000       A   M         4           NaN    NaN      4      4   
4505 245.000     NaN       A   M         4           NaN    NaN      4      4   
4506 250.000 234.000       A   M         4           NaN    NaN      4      4   
4507 245.000 209.000       A   F         4        29.100    NaN      3      3   

     Comments_1 Ager_2  AGE_2  FSP_2 Comments_2 Ager_3  AGE_3  FSP_3  \
4503        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4504        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4505        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4506        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   
4507        NaN    NaN   <NA>   <NA>        NaN    NaN   <NA>   <NA>   

     Comments_3 Envelop.Comments   DATETIME AGE_notes_1 FSP_notes_1  \
4503        NaN              NaN 1988-05-23           4           4   
4504        NaN              NaN 1988-05-23           4           4   
4505        NaN              NaN 1988-05-23           4           4   
4506        NaN              NaN 1988-05-23           4           4   
4507        NaN              NaN 1988-05-23           3           3   

     AGE_notes_2 FSP_notes_2 AGE_notes_3 FSP_notes_3 SITE_notes  SITE1 SITE2  \
4503         NaN         NaN         NaN         NaN         12     12  <NA>   
4504         NaN         NaN         NaN         NaN         12     12  <NA>   
4505         NaN         NaN         NaN         NaN         12     12  <NA>   
4506         NaN         NaN         NaN         NaN         12     12  <NA>   
4507         NaN         NaN         NaN         NaN         12     12  <NA>   

     SITE3  fish_length                                         remarks  \
4503  <NA>          253  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4504  <NA>          258  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4505  <NA>          245  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4506  <NA>          250  SITE_notes: 12; AGE_notes_1: 4; FSP_notes_1: 4   
4507  <NA>          245  SITE_notes: 12; AGE_notes_1: 3; FSP_notes_1: 3   

              id FLAG_SITE FLAG_AM_PM_PERIOD FLAG_SEX FLAG_MATURITY  \
4503  1988052312       NaN               NaN      NaN           NaN   
4504  1988052312       NaN               NaN      NaN           NaN   
4505  1988052312       NaN               NaN      NaN           NaN   
4506  1988052312       NaN               NaN      NaN           NaN   
4507  1988052312       NaN               NaN      NaN           NaN   

     FLAG_FSP_1 FLAG_FL_STD FLAG_FL_WET_FROZEN FLAG_WEIGHT_OUTLIER  \
4503        NaN         NaN                NaN                 NaN   
4504        NaN         NaN                NaN                 NaN   
4505        NaN         NaN                NaN                 NaN   
4506        NaN         NaN                NaN                 NaN   
4507        NaN         NaN                NaN                 NaN   

     FLAG_GONAD_OUTLIER FLAG_MULTIPLE_SAMPLE_POSSIBILITIES  \
4503                NaN                                NaN   
4504                NaN                                NaN   
4505                NaN                                NaN   
4506                NaN                                NaN   
4507                NaN                                NaN   

     FLAG_MISNUMBERED_FISH_DETAILS FL


1988 4 0 37


Empty DataFrame
Columns: [yy, mm, dd, Time, river, week, site, loc, period, wt_lbs, wt_kg, lgth, freq, Flbin, DATETIME, SITE1, SITE2, SITE3, id, length_bin_id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

Empty DataFrame
Columns: [YEAR, MM, DD, WEEK, SITE, PERIOD, CONDITION, FISH_NO, FL_WET, FL_FROZEN, FL_STD, WEIGHT, SPECIES, SEX, MATURITY, GONAD_WEIGHT, Ager_1, AGE_1, FSP_1, Comments_1, Ager_2, AGE_2, FSP_2, Comments_2, Ager_3, AGE_3, FSP_3, Comments_3, Envelop.Comments, DATETIME, AGE_notes_1, FSP_notes_1, AGE_notes_2, FSP_notes_2, AGE_notes_3, FSP_notes_3, SITE_notes, SITE1, SITE2, SITE3, fish_length, remarks, id, FLAG_SITE, FLAG_AM_PM_PERIOD, FLAG_SEX, FLAG_MATURITY, FLAG_FSP_1, FLAG_FL_STD, FLAG_FL_WET_FROZEN, FLAG_WEIGHT_OUTLIER, FLAG_GONAD_OUTLIER, FLAG_MULTIPLE_SAMPLE_POSSIBILITIES, FLAG_MISNUMBERED_FISH_DETAILS, FLAG_LEN_WT_RATIO_OUTLIER, FLAG_AM_PM_PERIOD_DISCREPANCIES, FLAG_NO_MATCHING_SAMPLE]
Index: []

### how do we flag and leave the potential to merge ghost/ambiguous samples?

In [9]:
id = 1990061312
date = id // 100
date

19900613

In [10]:
from datetime import datetime
datetime.strptime(str(id//100), "%Y%m%d").date()

datetime.date(1990, 6, 13)

In [11]:
id = 4990061312
date = id // 100
while date > 20240000:
    date -= 10000000
datetime.strptime(str(date), "%Y%m%d").date()

datetime.date(1990, 6, 13)

In [12]:
sum(df_LF.id.isna()), sum(df_FD.id.isna())

(0, 0)

In [13]:
id - 100*(id//100)

12

# When making ghost samples, will any of our ambiguous id match between fish details and length frequencies?

In [14]:
ambiguous_FD = set(df_FD[df_FD.id>2024000000].id)
ambiguous_LF = set(df_LF[df_LF.id>2024000000].id)
ambiguous_SD = set(df_SD[df_SD.id>2024000000].id)
# yes, we need to further disambiguate: maybe add 20/40 to months, this number would never occur naturally

In [15]:
# let's make a summary of ambiguous dates

df_A_FD = pd.DataFrame(ambiguous_FD, columns=['id_FD'])
df_A_FD['date'] = df_A_FD['id_FD'] // 100
while df_A_FD.date.max() > 20240000:
    df_A_FD.loc[df_A_FD.date>20240000, 'date'] -= 10000000
    
df_A_LF = pd.DataFrame(ambiguous_LF, columns=['id_LF'])
df_A_LF['date'] = df_A_LF['id_LF'] // 100
while df_A_LF.date.max() > 20240000:
    df_A_LF.loc[df_A_LF.date>20240000, 'date'] -= 10000000
    
df_A_SD = pd.DataFrame(ambiguous_SD, columns=['id_SD'])
df_A_SD['date'] = df_A_SD['id_SD'] // 100
while df_A_SD.date.max() > 20240000:
    df_A_SD.loc[df_A_SD.date>20240000, 'date'] -= 10000000

# group and split into columns
df_A_FD = df_A_FD.groupby('date')['id_FD'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_FD.columns = ['id_FD' for _ in range(df_A_FD.shape[1])]

df_A_LF = df_A_LF.groupby('date')['id_LF'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_LF.columns = ['id_LF' for _ in range(df_A_LF.shape[1])]

df_A_SD = df_A_SD.groupby('date')['id_SD'].apply(lambda x: pd.Series(x.values)).unstack()
df_A_SD.columns = ['id_SD' for _ in range(df_A_SD.shape[1])]

# create a dataframe summarising ambiguous data
df_AMBIGUOUS = pd.merge(
    df_A_FD,
    pd.merge(
        df_A_SD,
        df_A_LF,
        on='date', 
        how='outer'
    ),
    on='date', 
    how='outer'
).sort_index().astype('Int64')

df_AMBIGUOUS.loc[df_AMBIGUOUS['id_FD'].any(axis=1), 'FLAG_FD_AMBIGUOUS'] = True
df_AMBIGUOUS.loc[df_AMBIGUOUS['id_LF'].any(axis=1), 'FLAG_LF_AMBIGUOUS'] = True
df_AMBIGUOUS.loc[df_AMBIGUOUS['id_SD'].any(axis=1), 'FLAG_SD_AMBIGUOUS'] = True

df_AMBIGUOUS['FLAG_AMBIGUITY_OVERLAP'] = df_AMBIGUOUS.FLAG_FD_AMBIGUOUS.notnull()*1 + df_AMBIGUOUS.FLAG_LF_AMBIGUOUS.notnull()*1 + df_AMBIGUOUS.FLAG_SD_AMBIGUOUS.notnull()*1

In [16]:
# these are our problems if we are making ghost samples

df_AMBIGUOUS.loc[df_AMBIGUOUS.FLAG_AMBIGUITY_OVERLAP > 1].dropna(axis=1).T

date                      19980501    20000607    20000609    20010516  \
id_FD                   2998050105  3000060726  4000060926  3001051605   
id_LF                   3998050105  4000060726  3000060926  3001051605   
id_LF                   2998050105  3000060726  4000060926  4001051605   
FLAG_FD_AMBIGUOUS             True        True        True        True   
FLAG_LF_AMBIGUOUS             True        True        True        True   
FLAG_AMBIGUITY_OVERLAP           2           2           2           2   

date                      20090520    20090522    20100511    20100519  \
id_FD                   4009052026  3009052226  3010051126  3010051926   
id_LF                   3009052026  3009052226  3010051126  3010051926   
id_LF                   4009052026  4009052226  4010051126  4010051926   
FLAG_FD_AMBIGUOUS             True        True        True        True   
FLAG_LF_AMBIGUOUS             True        True        True        True   
FLAG_AMBIGUITY_OVERLAP           2           2           2           2   

date                      20100527  
id_FD                   3010052726  
id_LF                   3010052726  
id_LF                   4010052726  
FLAG_FD_AMBIGUOUS             True  
FLAG_LF_AMBIGUOUS             True  
FLAG_AMBIGUITY_OVERLAP           2

# Potential Sorting Issues? LF? FD?
* Yes, can't disambiguate without making guesses. Left flagged, and will be rejected on import.

In [17]:
for i in ambiguous_FD:
    display(df_FD[df_FD.id==i][['YEAR', 'MM', 'DD', 'SITE', 'PERIOD', 'FISH_NO', 'id']].head(1))

YEAR  MM  DD SITE PERIOD  FISH_NO          id
35924  1989   5  14   12     AM        1  2989051412

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36560  1989   5  14   12     PM        1  3989051412

YEAR  MM  DD SITE PERIOD  FISH_NO          id
35997  1993   5  29   52     AM        1  2993052952

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36730  2009   5  20   26     PM        1  4009052026

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36647  2000   6   9   26     PM        1  4000060926

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36110  2000   6   7   26     AM        1  3000060726

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36484  2014   5  30   47     AM        1  3014053047

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36327  2010   5  11   26     AM        1  3010051126

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36059  1998   5   1    5     PM        1  2998050105

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36607  2000   6   7   26     PM        1  4000060726

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36790  2010   5  11   26     PM        1  4010051126

YEAR  MM  DD      SITE PERIOD  FISH_NO          id
36882  2014   5  30  47 or 62     PM        1  4014053047

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36421  2010   5  29   25     PM        1  3010052925

YEAR  MM  DD SITE PERIOD  FISH_NO          id
35971  1993   5  27   37     AM        1  2993052737

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36264  2009   5  22   26     AM        1  3009052226

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36760  2009   5  22   26     PM        1  4009052226

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36033  1993   6   9   33     AM        1  2993060933

YEAR  MM  DD SITE PERIOD  FISH_NO          id
35878  1989   5  13   35     AM        1  2989051335

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36514  1989   5  13   35     PM        1  3989051335

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36452  2010   6  11   41     AM        1  3010061141

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36188  2001   5  16    5     PM        1  3001051605

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36355  2010   5  19   26     AM        1  3010051926

YEAR  MM  DD  SITE PERIOD  FISH_NO          id
36684  2001   5  16  5,26     AM        1  4001051605

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36817  2010   5  19   26     PM        1  4010051926

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36294  2010   5   5   41     PM        1  3010050541

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36392  2010   5  27   26     AM        1  3010052726

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36853  2010   5  27   26     PM        1  4010052726

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36234  2009   5  20   26     AM        1  3009052026

YEAR  MM  DD SITE PERIOD  FISH_NO          id
36151  2000   6   9   26     AM        1  3000060926

In [18]:
# CONFIRM
# some have am / pm that could be used to match with LF manually if that info is there as well???
for i in ambiguous_FD:
    print('\n', i)
    display(df_FD[df_FD.id==i][['YEAR', 'MM', 'DD', 'SITE', 'PERIOD', 'FISH_NO']].head())


 2989051412


YEAR  MM  DD SITE PERIOD  FISH_NO
35924  1989   5  14   12     AM        1
35925  1989   5  14   12     AM        2
35926  1989   5  14   12     AM        3
35927  1989   5  14   12     AM        4
35928  1989   5  14   12     AM        5


 3989051412


YEAR  MM  DD SITE PERIOD  FISH_NO
36560  1989   5  14   12     PM        1
36561  1989   5  14   12     PM        2
36562  1989   5  14   12     PM        3
36563  1989   5  14   12     PM        4
36564  1989   5  14   12     PM        5


 2993052952


YEAR  MM  DD SITE PERIOD  FISH_NO
35997  1993   5  29   52     AM        1
35998  1993   5  29   52     AM        2
35999  1993   5  29   52     AM        3
36000  1993   5  29   52     AM        4
36001  1993   5  29   52     AM        5


 4009052026


YEAR  MM  DD SITE PERIOD  FISH_NO
36730  2009   5  20   26     PM        1
36731  2009   5  20   26     PM        2
36732  2009   5  20   26     PM        3
36733  2009   5  20   26     PM        4
36734  2009   5  20   26     PM        5


 4000060926


YEAR  MM  DD SITE PERIOD  FISH_NO
36647  2000   6   9   26     PM        1
36648  2000   6   9   26     PM        2
36649  2000   6   9   26     PM        3
36650  2000   6   9   26     PM        4
36651  2000   6   9   26     PM        5


 3000060726


YEAR  MM  DD SITE PERIOD  FISH_NO
36110  2000   6   7   26     AM        1
36111  2000   6   7   26     AM        2
36112  2000   6   7   26     AM        3
36113  2000   6   7   26     AM        4
36114  2000   6   7   26     AM        5


 3014053047


YEAR  MM  DD SITE PERIOD  FISH_NO
36484  2014   5  30   47     AM        1
36485  2014   5  30   47     AM        2
36486  2014   5  30   47     AM        3
36487  2014   5  30   47     AM        4
36488  2014   5  30   47     AM        5


 3010051126


YEAR  MM  DD SITE PERIOD  FISH_NO
36327  2010   5  11   26     AM        1
36328  2010   5  11   26     AM        2
36329  2010   5  11   26     AM        3
36330  2010   5  11   26     AM        4
36331  2010   5  11   26     AM        5


 2998050105


YEAR  MM  DD SITE PERIOD  FISH_NO
36059  1998   5   1    5     PM        1
36060  1998   5   1    5     PM        1
36061  1998   5   1    5     PM        2
36062  1998   5   1    5     PM        2
36063  1998   5   1    5     PM        3


 4000060726


YEAR  MM  DD SITE PERIOD  FISH_NO
36607  2000   6   7   26     PM        1
36608  2000   6   7   26     PM        2
36609  2000   6   7   26     PM        3
36610  2000   6   7   26     PM        4
36611  2000   6   7   26     PM        5


 4010051126


YEAR  MM  DD SITE PERIOD  FISH_NO
36790  2010   5  11   26     PM        1
36791  2010   5  11   26     PM        2
36792  2010   5  11   26     PM        3
36793  2010   5  11   26     PM        4
36794  2010   5  11   26     PM        5


 4014053047


YEAR  MM  DD      SITE PERIOD  FISH_NO
36882  2014   5  30  47 or 62     PM        1
36883  2014   5  30  47 or 62     PM        2
36884  2014   5  30  47 or 62     PM        3
36885  2014   5  30  47 or 62     PM        4
36886  2014   5  30  47 or 62     PM        5


 3010052925


YEAR  MM  DD SITE PERIOD  FISH_NO
36421  2010   5  29   25     PM        1
36422  2010   5  29   25     PM        2
36423  2010   5  29   25     PM        3
36424  2010   5  29   25     PM        4
36425  2010   5  29   25     PM        5


 2993052737


YEAR  MM  DD SITE PERIOD  FISH_NO
35971  1993   5  27   37     AM        1
35972  1993   5  27   37     AM        2
35973  1993   5  27   37     AM        3
35974  1993   5  27   37     AM        4
35975  1993   5  27   37     AM        5


 3009052226


YEAR  MM  DD SITE PERIOD  FISH_NO
36264  2009   5  22   26     AM        1
36265  2009   5  22   26     AM        2
36266  2009   5  22   26     AM        3
36267  2009   5  22   26     AM        4
36268  2009   5  22   26     AM        5


 4009052226


YEAR  MM  DD SITE PERIOD  FISH_NO
36760  2009   5  22   26     PM        1
36761  2009   5  22   26     PM        2
36762  2009   5  22   26     PM        3
36763  2009   5  22   26     PM        4
36764  2009   5  22   26     PM        5


 2993060933


YEAR  MM  DD SITE PERIOD  FISH_NO
36033  1993   6   9   33     AM        1
36034  1993   6   9   33     AM        2
36035  1993   6   9   33     AM        3
36036  1993   6   9   33     AM        4
36037  1993   6   9   33     AM        5


 2989051335


YEAR  MM  DD SITE PERIOD  FISH_NO
35878  1989   5  13   35     AM        1
35879  1989   5  13   35     AM        2
35880  1989   5  13   35     AM        3
35881  1989   5  13   35     AM        4
35882  1989   5  13   35     AM        5


 3989051335


YEAR  MM  DD SITE PERIOD  FISH_NO
36514  1989   5  13   35     PM        1
36515  1989   5  13   35     PM        2
36516  1989   5  13   35     PM        3
36517  1989   5  13   35     PM        4
36518  1989   5  13   35     PM        5


 3010061141


YEAR  MM  DD SITE PERIOD  FISH_NO
36452  2010   6  11   41     AM        1
36453  2010   6  11   41     AM        2
36454  2010   6  11   41     AM        3
36455  2010   6  11   41     AM        4
36456  2010   6  11   41     AM        5


 3001051605


YEAR  MM  DD SITE PERIOD  FISH_NO
36188  2001   5  16    5     PM        1
36189  2001   5  16    5     PM        2
36190  2001   5  16    5     PM        3
36191  2001   5  16    5     PM        4
36192  2001   5  16    5     PM        5


 3010051926


YEAR  MM  DD SITE PERIOD  FISH_NO
36355  2010   5  19   26     AM        1
36356  2010   5  19   26     AM        2
36357  2010   5  19   26     AM        3
36358  2010   5  19   26     AM        4
36359  2010   5  19   26     AM        5


 4001051605


YEAR  MM  DD  SITE PERIOD  FISH_NO
36684  2001   5  16  5,26     AM        1
36685  2001   5  16  5,26     AM        2
36686  2001   5  16  5,26     AM        3
36687  2001   5  16  5,26     AM        4
36688  2001   5  16  5,26     AM        5


 4010051926


YEAR  MM  DD SITE PERIOD  FISH_NO
36817  2010   5  19   26     PM        1
36818  2010   5  19   26     PM        2
36819  2010   5  19   26     PM        3
36820  2010   5  19   26     PM        4
36821  2010   5  19   26     PM        5


 3010050541


YEAR  MM  DD SITE PERIOD  FISH_NO
36294  2010   5   5   41     PM        1
36295  2010   5   5   41     PM        2
36296  2010   5   5   41     PM        3
36297  2010   5   5   41     PM        4
36298  2010   5   5   41     PM        5


 3010052726


YEAR  MM  DD SITE PERIOD  FISH_NO
36392  2010   5  27   26     AM        1
36393  2010   5  27   26     AM        2
36394  2010   5  27   26     AM        3
36395  2010   5  27   26     AM        4
36396  2010   5  27   26     AM        5


 4010052726


YEAR  MM  DD SITE PERIOD  FISH_NO
36853  2010   5  27   26     PM        1
36854  2010   5  27   26     PM        2
36855  2010   5  27   26     PM        3
36856  2010   5  27   26     PM        4
36857  2010   5  27   26     PM        5


 3009052026


YEAR  MM  DD SITE PERIOD  FISH_NO
36234  2009   5  20   26     AM        1
36235  2009   5  20   26     AM        2
36236  2009   5  20   26     AM        3
36237  2009   5  20   26     AM        4
36238  2009   5  20   26     AM        5


 3000060926


YEAR  MM  DD SITE PERIOD  FISH_NO
36151  2000   6   9   26     AM        1
36152  2000   6   9   26     AM        2
36153  2000   6   9   26     AM        3
36154  2000   6   9   26     AM        4
36155  2000   6   9   26     AM        5

In [19]:
# check ambiguous AM PM for LF and FD and look for extra matches (NOTE, none match samples/logbook)
# [x for x in df_AMBIGUOUS[['id_LF']].values.T.ravel().tolist() if not pd.isna(x)]

period_check = pd.merge(
    df_FD.loc[df_FD.id.isin([x for x in df_AMBIGUOUS[['id_FD']].values.T.ravel().tolist() if not pd.isna(x)])].groupby(['id', 'PERIOD']).count().reset_index()[['id', 'PERIOD']],
    df_LF.loc[df_LF.id.isin([x for x in df_AMBIGUOUS[['id_LF']].values.T.ravel().tolist() if not pd.isna(x)])].groupby(['id', 'period']).count().reset_index()[['id', 'period']],
    on='id',
    how='outer'
)

period_check['id'] -= 1000000000
period_check.loc[period_check.id > 2024000000, 'id'] -= 1000000000

period_check = period_check.sort_values(['id', 'PERIOD']).reset_index(drop=True)

In [20]:
id_period_ambiguity = list(period_check.dropna()[period_check.dropna().duplicated('id', keep=False)].id.unique())

check_ids = [x+1000000000 for x in id_period_ambiguity] + [x+2000000000 for x in id_period_ambiguity]

In [21]:
df_LF[df_LF.id.isin(check_ids)].groupby(['id', 'period']).first().sort_values('DATETIME')

yy  mm  dd     Time        river  week  site    loc  \
id         period                                                          
3000060726 AM      2000   6   7      NaN  SW MARGAREE     6    26  LOWER   
4000060726 PM      2000   6   7      NaN  SW MARGAREE     6    26  LOWER   
3000060926 PM      2000   6   9      NaN  SW MARGAREE     6    26  LOWER   
4000060926 AM      2000   6   9      NaN  SW MARGAREE     6    26  LOWER   
3001051605 AM      2001   5  16      NaN  SW MARGAREE     3  5,26  LOWER   
4001051605 PM      2001   5  16      NaN  SW MARGAREE     3     5  LOWER   
3009052026 AM      2009   5  20 1255.000  SW MARGAREE     4    26  LOWER   
4009052026 PM      2009   5  20 1405.000  SW MARGAREE     4    26  LOWER   
3009052226 AM      2009   5  22 1310.000  SW MARGAREE     4    26  LOWER   
4009052226 PM      2009   5  22 1455.000  SW MARGAREE     4    26  LOWER   
3010051126 AM      2010   5  11 1450.000  SW MARGAREE     3    26  LOWER   
4010051126 PM      2010   5  11 1540.000  SW MARGAREE     3    26  LOWER   
3010051926 AM      2010   5  19 1345.000  SW MARGAREE     4    26  LOWER   
4010051926 PM      2010   5  19 1630.000  SW MARGAREE     4    26  LOWER   
3010052726 AM      2010   5  27  840.000  SW MARGAREE     5    26  LOWER   
4010052726 PM      2010   5  27 1500.000  SW MARGAREE     5    26  LOWER   

                   wt_lbs  wt_kg  lgth  freq  Flbin            DATETIME SITE1  \
id         period                                                               
3000060726 AM     158.000 71.700   215     1    215 2000-06-07 00:00:00    26   
4000060726 PM     158.000 71.700   215     3    215 2000-06-07 00:00:00    26   
3000060926 PM     141.000 64.000   215     3    215 2000-06-09 00:00:00    26   
4000060926 AM     146.000 66.200   215     1    215 2000-06-09 00:00:00    26   
3001051605 AM     140.000 63.500   220     1    220 2001-05-16 00:00:00     5   
4001051605 PM     154.000 69.900   225     2    225 2001-05-16 00:00:00     5   
3009052026 AM     139.000 63.000   220     1    220 2009-05-20 12:55:00    26   
4009052026 PM      96.000 43.500   225     1    225 2009-05-20 14:05:00    26   
3009052226 AM     128.000 58.100   225     2    225 2009-05-22 13:10:00    26   
4009052226 PM     130.000 59.000   225     3    225 2009-05-22 14:55:00    26   
3010051126 AM      43.000 19.500   230     3    230 2010-05-11 14:50:00    26   
4010051126 PM      36.000 16.300   235     1    235 2010-05-11 15:40:00    26   
3010051926 AM     129.000 58.500   225     2    225 2010-05-19 13:45:00    26   
4010051926 PM     100.000 45.400   220     1    220 2010-05-19 16:30:00    26   
3010052726 AM     117.000 53.100   225     1    225 2010-05-27 08:40:00    26   
4010052726 PM     121.000 54.900   230    19    230 2010-05-27 15:00:00    26   

                  SITE2 SITE3  length_bin_id FLAG_SITE FLAG_AM_PM_PERIOD  \
id         period                                                          
3000060726 AM      None  None         21.500      None              None   
4000060726 PM      None  None         21.500      None              None   
3000060926 PM      None  None         21.500      None              None   
4000060926 AM      None  None         21.500      None              None   
3001051605 AM        26  None         22.000      True              None   
4001051605 PM      None  None         22.500      None              None   
3009052026 AM      None  None         22.000      None              None   
4009052026 PM      None  None         22.500      None              None   
3009052226 AM      None  None         22.500      None              None   
4009052226 PM      None  None         22.500      None              None   
3010051126 AM      None  None         23.000      None              None   
4010051126 PM      None  None         23.500      None              None   
3010051926 AM      None  None         22.500      None              None   
4010051926 PM      None  None         22.000      Non

In [22]:
df_FD[df_FD.id.isin(check_ids)].groupby(['id', 'PERIOD']).first().sort_values('DATETIME')

YEAR  MM  DD  WEEK  SITE CONDITION  FISH_NO  FL_WET  \
id         PERIOD                                                        
3000060726 AM      2000   6   7     6    26    Frozen        1    <NA>   
           PM      2000   6   7     6    26     Fresh       37     242   
4000060726 PM      2000   6   7     6    26     Fresh        1     252   
3000060926 AM      2000   6   9     6    26     Fresh        1     239   
           PM      2000   6   9     6    26    Frozen       35    <NA>   
4000060926 PM      2000   6   9     6    26    Frozen        1    <NA>   
3001051605 AM      2001   5  16     3  5,26    Frozen       46    <NA>   
           PM      2001   5  16     3     5    Frozen        1    <NA>   
4001051605 AM      2001   5  16     3  5,26    Frozen        1    <NA>   
3009052026 AM      2009   5  20     4    26     Fresh        1     245   
           PM      2009   5  20     4    26    Frozen       30    <NA>   
4009052026 PM      2009   5  20     4    26    Frozen        1    <NA>   
3009052226 AM      2009   5  22     4    26     Fresh        1     232   
4009052226 AM      2009   5  22     4    26     Fresh       31     247   
           PM      2009   5  22     4    26    Frozen        1    <NA>   
4010051126 PM      2010   5  11     3    26    Frozen        1    <NA>   
3010051126 AM      2010   5  11     3    26    Frozen        1    <NA>   
4010051126 AM      2010   5  11     3    26    Frozen       28    <NA>   
3010051926 AM      2010   5  19     4    26    Frozen        1    <NA>   
4010051926 PM      2010   5  19     4    26    Frozen        1    <NA>   
3010052726 AM      2010   5  27     5    26     Fresh        1     235   
4010052726 PM      2010   5  27     5    26    Frozen        1    <NA>   

                   FL_FROZEN  FL_STD  WEIGHT SPECIES SEX  MATURITY  \
id         PERIOD                                                    
3000060726 AM            250 258.000 220.000       B   M         4   
           PM           <NA> 242.000 198.000       B   M         4   
4000060726 PM           <NA> 252.000 226.000       B   F         4   
3000060926 AM           <NA> 239.000 168.000       B   M         4   
           PM            228 236.000 154.800       A   M         4   
4000060926 PM            266 274.000 284.500       A   F         4   
3001051605 AM            217 225.000 128.400       A   M         4   
           PM            276 285.000 312.600       A   F         4   
4001051605 AM            271 279.000 288.000       A   M         4   
3009052026 AM            269 245.000 202.500       A   M         4   
           PM            265 273.000 257.800       A   F         4   
4009052026 PM            243 251.000 200.800       A   M         4   
3009052226 AM           <NA> 232.000 156.000       A   M         4   
4009052226 AM           <NA> 247.000 225.400       A   F         4   
           PM            226 234.000 160.200       A   M         4   
4010051126 PM            242 250.000 172.100       A   M         4   
3010051126 AM            251 259.000 215.500       A   M         4   
4010051126 AM            260 268.000 234.000       A   M         4   
3010051926 AM            273 281.000 271.100       A   F         4   
4010051926 PM            235 243.000 184.100       A   F         4   
3010052726 AM           <NA> 235.000 163.100       A   M         4   
4010052726 PM            239 247.000 191.700       A   M         4   

                   GONAD_WEIGHT Ager_1  AGE_1  FSP_1 Comments_1 Ager_2  AGE_2  \
id         PERIOD                                                               
3000060726 AM            32.100   None      3      9       None   None   <NA>   
           PM            52.500   None      3      9       None   None   <NA>   
4000060726 PM            28.800   None      3      9       None   None   <NA>   
3000060926 AM            32.700   None      3      0       None   None   <NA>   
           PM            35.200   None   <NA>      9       None   None   <N

4 ambiguous samples may be linked:

    3010051926	AM
    3010051926	AM
    
    4010051926	PM
    4010051926	PM
    
    3010052726	AM
    3010052726	AM
    
    4010052726	PM
    4010052726	PM
    
Note: these are matched together correctly.

* Therefore, 

        if id.isin([3010051926, 4010051926, 3010052726, 4010052726]):
            don't add anything to months to disambiguate 
            # these id will match up correctly and link to the same ghost sample

* Other id from this set do not match unambiguously using period. There are some inconsistencies.
* In fact, period may not be entirely useful because of the inconsistencies with within ambiguous sets of fish details.

# More Sorting Issue Calcs and Checks

In [23]:
# loop through duplicated ids and look at these by inspection - make sure they all look doubled (except 2018061364)
# the only one that looks like a typo is 2018061364 (fixed at "CONFIRM: fix typo at 10851" in noteboook 10.1)

# from notebook 10.1
duplicated_ids_LF = [1998050105, 2000060726, 2000060926, 2000061538, 2001051605, 2003052326,
                     2009052026, 2009052226, 2010051126, 2010051926, 2010052726, 2018061364]

temp = df_LF.copy()
temp.loc[temp.id>2024000000, 'id'] -= 1000000000
temp.loc[temp.id>2024000000, 'id'] -= 1000000000

for duplicate_id in duplicated_ids_LF:
    idx = temp[temp.id == duplicate_id].index.values[0]
    
    before = 0
    after = temp[temp.id == duplicate_id].shape[0]-1
    
    print('\n', temp[temp.id == duplicate_id]['id'].values[0])
    display(temp.loc[idx-before:idx+after][['id', 'DATETIME', 'site', 'period', 'lgth', 'wt_lbs']])



 1998050105


id   DATETIME site period  lgth  wt_lbs
3102  1998050105 1998-05-01    5     PM   235  10.000
3103  1998050105 1998-05-01    5     PM   240  10.000
3104  1998050105 1998-05-01    5     PM   245  10.000
3105  1998050105 1998-05-01    5     PM   250  10.000
3106  1998050105 1998-05-01    5     PM   255  10.000
3107  1998050105 1998-05-01    5     PM   260  10.000
3108  1998050105 1998-05-01    5     PM   265  10.000
3109  1998050105 1998-05-01    5     PM   270  10.000
3110  1998050105 1998-05-01    5     PM   275  10.000
3111  1998050105 1998-05-01    5     PM   280  10.000
3112  1998050105 1998-05-01    5     AM   230 132.000
3113  1998050105 1998-05-01    5     AM   235 132.000
3114  1998050105 1998-05-01    5     AM   240 132.000
3115  1998050105 1998-05-01    5     AM   245 132.000
3116  1998050105 1998-05-01    5     AM   250 132.000
3117  1998050105 1998-05-01    5     AM   255 132.000
3118  1998050105 1998-05-01    5     AM   260 132.000
3119  1998050105 1998-05-01    5     AM   265 132.000
3120  1998050105 1998-05-01    5     AM   270 132.000
3121  1998050105 1998-05-01    5     AM   275 132.000
3122  1998050105 1998-05-01    5     AM   280 132.000


 2000060726


id   DATETIME site period  lgth  wt_lbs
4156  2000060726 2000-06-07   26     AM   215 158.000
4157  2000060726 2000-06-07   26     AM   220 158.000
4158  2000060726 2000-06-07   26     AM   225 158.000
4159  2000060726 2000-06-07   26     AM   230 158.000
4160  2000060726 2000-06-07   26     AM   235 158.000
4161  2000060726 2000-06-07   26     AM   240 158.000
4162  2000060726 2000-06-07   26     AM   245 158.000
4163  2000060726 2000-06-07   26     AM   250 158.000
4164  2000060726 2000-06-07   26     AM   255 158.000
4165  2000060726 2000-06-07   26     AM   260 158.000
4166  2000060726 2000-06-07   26     AM   265 158.000
4167  2000060726 2000-06-07   26     AM   270 158.000
4168  2000060726 2000-06-07   26     AM   275 158.000
4169  2000060726 2000-06-07   26     AM   280 158.000
4170  2000060726 2000-06-07   26     AM   285 158.000
4171  2000060726 2000-06-07   26     PM   215 158.000
4172  2000060726 2000-06-07   26     PM   220 158.000
4173  2000060726 2000-06-07   26     PM   225 158.000
4174  2000060726 2000-06-07   26     PM   230 158.000
4175  2000060726 2000-06-07   26     PM   235 158.000
4176  2000060726 2000-06-07   26     PM   240 158.000
4177  2000060726 2000-06-07   26     PM   245 158.000
4178  2000060726 2000-06-07   26     PM   250 158.000
4179  2000060726 2000-06-07   26     PM   255 158.000
4180  2000060726 2000-06-07   26     PM   260 158.000
4181  2000060726 2000-06-07   26     PM   265 158.000
4182  2000060726 2000-06-07   26     PM   270 158.000
4183  2000060726 2000-06-07   26     PM   275 158.000
4184  2000060726 2000-06-07   26     PM   280 158.000
4185  2000060726 2000-06-07   26     PM   285 158.000


 2000060926


id   DATETIME site period  lgth  wt_lbs
4186  2000060926 2000-06-09   26     PM   215 141.000
4187  2000060926 2000-06-09   26     PM   220 141.000
4188  2000060926 2000-06-09   26     PM   225 141.000
4189  2000060926 2000-06-09   26     PM   230 141.000
4190  2000060926 2000-06-09   26     PM   235 141.000
4191  2000060926 2000-06-09   26     PM   240 141.000
4192  2000060926 2000-06-09   26     PM   245 141.000
4193  2000060926 2000-06-09   26     PM   250 141.000
4194  2000060926 2000-06-09   26     PM   255 141.000
4195  2000060926 2000-06-09   26     PM   260 141.000
4196  2000060926 2000-06-09   26     PM   265 141.000
4197  2000060926 2000-06-09   26     PM   270 141.000
4198  2000060926 2000-06-09   26     PM   275 141.000
4199  2000060926 2000-06-09   26     PM   280 141.000
4200  2000060926 2000-06-09   26     AM   215 146.000
4201  2000060926 2000-06-09   26     AM   220 146.000
4202  2000060926 2000-06-09   26     AM   225 146.000
4203  2000060926 2000-06-09   26     AM   230 146.000
4204  2000060926 2000-06-09   26     AM   235 146.000
4205  2000060926 2000-06-09   26     AM   240 146.000
4206  2000060926 2000-06-09   26     AM   245 146.000
4207  2000060926 2000-06-09   26     AM   250 146.000
4208  2000060926 2000-06-09   26     AM   255 146.000
4209  2000060926 2000-06-09   26     AM   260 146.000
4210  2000060926 2000-06-09   26     AM   265 146.000
4211  2000060926 2000-06-09   26     AM   270 146.000
4212  2000060926 2000-06-09   26     AM   275 146.000


 2000061538


id   DATETIME site period  lgth  wt_lbs
4239  2000061538 2000-06-15   38     PM   205 139.000
4240  2000061538 2000-06-15   38     PM   210 139.000
4241  2000061538 2000-06-15   38     PM   215 139.000
4242  2000061538 2000-06-15   38     PM   220 139.000
4243  2000061538 2000-06-15   38     PM   225 139.000
4244  2000061538 2000-06-15   38     PM   230 139.000
4245  2000061538 2000-06-15   38     PM   235 139.000
4246  2000061538 2000-06-15   38     PM   240 139.000
4247  2000061538 2000-06-15   38     PM   245 139.000
4248  2000061538 2000-06-15   38     PM   250 139.000
4249  2000061538 2000-06-15   38     PM   255 139.000
4250  2000061538 2000-06-15   38     PM   260 139.000
4251  2000061538 2000-06-15   38     PM   265 139.000
4252  2000061538 2000-06-15   38     PM   270 139.000
4253  2000061538 2000-06-15   38     PM   275 139.000
4254  2000061538 2000-06-15   38     AM   220  57.000
4255  2000061538 2000-06-15   38     AM   225  57.000
4256  2000061538 2000-06-15   38     AM   230  57.000
4257  2000061538 2000-06-15   38     AM   235  57.000
4258  2000061538 2000-06-15   38     AM   240  57.000
4259  2000061538 2000-06-15   38     AM   245  57.000
4260  2000061538 2000-06-15   38     AM   250  57.000
4261  2000061538 2000-06-15   38     AM   255  57.000
4262  2000061538 2000-06-15   38     AM   260  57.000
4263  2000061538 2000-06-15   38     AM   265  57.000
4264  2000061538 2000-06-15   38     AM   270  57.000
4265  2000061538 2000-06-15   38     AM   275  57.000


 2001051605


id   DATETIME  site period  lgth  wt_lbs
4266  2001051605 2001-05-16  5,26     AM   220 140.000
4267  2001051605 2001-05-16  5,26     AM   225 140.000
4268  2001051605 2001-05-16  5,26     AM   230 140.000
4269  2001051605 2001-05-16  5,26     AM   235 140.000
4270  2001051605 2001-05-16  5,26     AM   240 140.000
4271  2001051605 2001-05-16  5,26     AM   245 140.000
4272  2001051605 2001-05-16  5,26     AM   250 140.000
4273  2001051605 2001-05-16  5,26     AM   255 140.000
4274  2001051605 2001-05-16  5,26     AM   260 140.000
4275  2001051605 2001-05-16  5,26     AM   265 140.000
4276  2001051605 2001-05-16  5,26     AM   270 140.000
4277  2001051605 2001-05-16  5,26     AM   275 140.000
4278  2001051605 2001-05-16  5,26     AM   280 140.000
4279  2001051605 2001-05-16  5,26     AM   285 140.000
4280  2001051605 2001-05-16  5,26     AM   295 140.000
4281  2001051605 2001-05-16     5     PM   225 154.000
4282  2001051605 2001-05-16     5     PM   230 154.000
4283  2001051605 2001-05-16     5     PM   235 154.000
4284  2001051605 2001-05-16     5     PM   240 154.000
4285  2001051605 2001-05-16     5     PM   245 154.000
4286  2001051605 2001-05-16     5     PM   250 154.000
4287  2001051605 2001-05-16     5     PM   255 154.000
4288  2001051605 2001-05-16     5     PM   260 154.000
4289  2001051605 2001-05-16     5     PM   265 154.000
4290  2001051605 2001-05-16     5     PM   270 154.000
4291  2001051605 2001-05-16     5     PM   275 154.000
4292  2001051605 2001-05-16     5     PM   280 154.000
4293  2001051605 2001-05-16     5     PM   285 154.000
4294  2001051605 2001-05-16     5     PM   290 154.000
4295  2001051605 2001-05-16     5     PM   295 154.000


 2003052326


id   DATETIME site period  lgth  wt_lbs
5184  2003052326 2003-05-23   26     AM   230  30.000
5185  2003052326 2003-05-23   26     AM   235  30.000
5186  2003052326 2003-05-23   26     AM   240  30.000
5187  2003052326 2003-05-23   26     AM   245  30.000
5188  2003052326 2003-05-23   26     AM   250  30.000
5189  2003052326 2003-05-23   26     AM   255  30.000
5190  2003052326 2003-05-23   26     AM   260  30.000
5191  2003052326 2003-05-23   26     AM   265  30.000
5192  2003052326 2003-05-23   26     AM   270  30.000
5193  2003052326 2003-05-23   26     AM   275  30.000
5194  2003052326 2003-05-23   26     PM   225  47.000
5195  2003052326 2003-05-23   26     PM   230  47.000
5196  2003052326 2003-05-23   26     PM   235  47.000
5197  2003052326 2003-05-23   26     PM   240  47.000
5198  2003052326 2003-05-23   26     PM   245  47.000
5199  2003052326 2003-05-23   26     PM   250  47.000
5200  2003052326 2003-05-23   26     PM   255  47.000
5201  2003052326 2003-05-23   26     PM   260  47.000
5202  2003052326 2003-05-23   26     PM   265  47.000
5203  2003052326 2003-05-23   26     PM   270  47.000
5204  2003052326 2003-05-23   26     PM   275  47.000
5205  2003052326 2003-05-23   26     PM   280  47.000


 2009052026


id            DATETIME site period  lgth  wt_lbs
6332  2009052026 2009-05-20 12:55:00   26     AM   220 139.000
6333  2009052026 2009-05-20 12:55:00   26     AM   225 139.000
6334  2009052026 2009-05-20 12:55:00   26     AM   230 139.000
6335  2009052026 2009-05-20 12:55:00   26     AM   235 139.000
6336  2009052026 2009-05-20 12:55:00   26     AM   240 139.000
6337  2009052026 2009-05-20 12:55:00   26     AM   245 139.000
6338  2009052026 2009-05-20 12:55:00   26     AM   250 139.000
6339  2009052026 2009-05-20 12:55:00   26     AM   255 139.000
6340  2009052026 2009-05-20 12:55:00   26     AM   260 139.000
6341  2009052026 2009-05-20 12:55:00   26     AM   265 139.000
6342  2009052026 2009-05-20 12:55:00   26     AM   270 139.000
6343  2009052026 2009-05-20 14:05:00   26     PM   225  96.000
6344  2009052026 2009-05-20 14:05:00   26     PM   230  96.000
6345  2009052026 2009-05-20 14:05:00   26     PM   235  96.000
6346  2009052026 2009-05-20 14:05:00   26     PM   240  96.000
6347  2009052026 2009-05-20 14:05:00   26     PM   245  96.000
6348  2009052026 2009-05-20 14:05:00   26     PM   250  96.000
6349  2009052026 2009-05-20 14:05:00   26     PM   255  96.000
6350  2009052026 2009-05-20 14:05:00   26     PM   260  96.000
6351  2009052026 2009-05-20 14:05:00   26     PM   265  96.000
6352  2009052026 2009-05-20 14:05:00   26     PM   270  96.000
6353  2009052026 2009-05-20 14:05:00   26     PM   275  96.000
6354  2009052026 2009-05-20 14:05:00   26     PM   280  96.000


 2009052226


id            DATETIME site period  lgth  wt_lbs
6368  2009052226 2009-05-22 13:10:00   26     AM   225 128.000
6369  2009052226 2009-05-22 13:10:00   26     AM   230 128.000
6370  2009052226 2009-05-22 13:10:00   26     AM   235 128.000
6371  2009052226 2009-05-22 13:10:00   26     AM   240 128.000
6372  2009052226 2009-05-22 13:10:00   26     AM   245 128.000
6373  2009052226 2009-05-22 13:10:00   26     AM   250 128.000
6374  2009052226 2009-05-22 13:10:00   26     AM   255 128.000
6375  2009052226 2009-05-22 13:10:00   26     AM   260 128.000
6376  2009052226 2009-05-22 13:10:00   26     AM   265 128.000
6377  2009052226 2009-05-22 13:10:00   26     AM   270 128.000
6378  2009052226 2009-05-22 13:10:00   26     AM   275 128.000
6379  2009052226 2009-05-22 14:55:00   26     PM   225 130.000
6380  2009052226 2009-05-22 14:55:00   26     PM   230 130.000
6381  2009052226 2009-05-22 14:55:00   26     PM   235 130.000
6382  2009052226 2009-05-22 14:55:00   26     PM   240 130.000
6383  2009052226 2009-05-22 14:55:00   26     PM   245 130.000
6384  2009052226 2009-05-22 14:55:00   26     PM   250 130.000
6385  2009052226 2009-05-22 14:55:00   26     PM   255 130.000
6386  2009052226 2009-05-22 14:55:00   26     PM   260 130.000
6387  2009052226 2009-05-22 14:55:00   26     PM   265 130.000
6388  2009052226 2009-05-22 14:55:00   26     PM   270 130.000


 2010051126


id            DATETIME site period  lgth  wt_lbs
6648  2010051126 2010-05-11 14:50:00   26     AM   230  43.000
6649  2010051126 2010-05-11 14:50:00   26     AM   235  43.000
6650  2010051126 2010-05-11 14:50:00   26     AM   240  43.000
6651  2010051126 2010-05-11 14:50:00   26     AM   245  43.000
6652  2010051126 2010-05-11 14:50:00   26     AM   250  43.000
6653  2010051126 2010-05-11 14:50:00   26     AM   255  43.000
6654  2010051126 2010-05-11 14:50:00   26     AM   260  43.000
6655  2010051126 2010-05-11 14:50:00   26     AM   265  43.000
6656  2010051126 2010-05-11 14:50:00   26     AM   270  43.000
6657  2010051126 2010-05-11 14:50:00   26     AM   275  43.000
6658  2010051126 2010-05-11 14:50:00   26     AM   280  43.000
6659  2010051126 2010-05-11 15:40:00   26     PM   235  36.000
6660  2010051126 2010-05-11 15:40:00   26     PM   240  36.000
6661  2010051126 2010-05-11 15:40:00   26     PM   245  36.000
6662  2010051126 2010-05-11 15:40:00   26     PM   250  36.000
6663  2010051126 2010-05-11 15:40:00   26     PM   255  36.000
6664  2010051126 2010-05-11 15:40:00   26     PM   260  36.000
6665  2010051126 2010-05-11 15:40:00   26     PM   265  36.000
6666  2010051126 2010-05-11 15:40:00   26     PM   270  36.000
6667  2010051126 2010-05-11 15:40:00   26     PM   275  36.000
6668  2010051126 2010-05-11 15:40:00   26     PM   280  36.000
6669  2010051126 2010-05-11 15:40:00   26     PM   285  36.000


 2010051926


id            DATETIME site period  lgth  wt_lbs
6719  2010051926 2010-05-19 13:45:00   26     AM   225 129.000
6720  2010051926 2010-05-19 13:45:00   26     AM   230 129.000
6721  2010051926 2010-05-19 13:45:00   26     AM   235 129.000
6722  2010051926 2010-05-19 13:45:00   26     AM   240 129.000
6723  2010051926 2010-05-19 13:45:00   26     AM   245 129.000
6724  2010051926 2010-05-19 13:45:00   26     AM   250 129.000
6725  2010051926 2010-05-19 13:45:00   26     AM   255 129.000
6726  2010051926 2010-05-19 13:45:00   26     AM   260 129.000
6727  2010051926 2010-05-19 13:45:00   26     AM   265 129.000
6728  2010051926 2010-05-19 13:45:00   26     AM   270 129.000
6729  2010051926 2010-05-19 13:45:00   26     AM   275 129.000
6730  2010051926 2010-05-19 13:45:00   26     AM   280 129.000
6731  2010051926 2010-05-19 13:45:00   26     AM   285 129.000
6732  2010051926 2010-05-19 16:30:00   26     PM   220 100.000
6733  2010051926 2010-05-19 16:30:00   26     PM   225 100.000
6734  2010051926 2010-05-19 16:30:00   26     PM   230 100.000
6735  2010051926 2010-05-19 16:30:00   26     PM   235 100.000
6736  2010051926 2010-05-19 16:30:00   26     PM   240 100.000
6737  2010051926 2010-05-19 16:30:00   26     PM   245 100.000
6738  2010051926 2010-05-19 16:30:00   26     PM   250 100.000
6739  2010051926 2010-05-19 16:30:00   26     PM   255 100.000
6740  2010051926 2010-05-19 16:30:00   26     PM   260 100.000
6741  2010051926 2010-05-19 16:30:00   26     PM   265 100.000
6742  2010051926 2010-05-19 16:30:00   26     PM   270 100.000
6743  2010051926 2010-05-19 16:30:00   26     PM   275 100.000
6744  2010051926 2010-05-19 16:30:00   26     PM   280 100.000


 2010052726


id            DATETIME site period  lgth  wt_lbs
6846  2010052726 2010-05-27 08:40:00   26     AM   225 117.000
6847  2010052726 2010-05-27 08:40:00   26     AM   230 117.000
6848  2010052726 2010-05-27 08:40:00   26     AM   235 117.000
6849  2010052726 2010-05-27 08:40:00   26     AM   240 117.000
6850  2010052726 2010-05-27 08:40:00   26     AM   245 117.000
6851  2010052726 2010-05-27 08:40:00   26     AM   250 117.000
6852  2010052726 2010-05-27 08:40:00   26     AM   255 117.000
6853  2010052726 2010-05-27 08:40:00   26     AM   260 117.000
6854  2010052726 2010-05-27 08:40:00   26     AM   265 117.000
6855  2010052726 2010-05-27 08:40:00   26     AM   270 117.000
6856  2010052726 2010-05-27 08:40:00   26     AM   275 117.000
6857  2010052726 2010-05-27 15:00:00   26     PM   230 121.000
6858  2010052726 2010-05-27 15:00:00   26     PM   235 121.000
6859  2010052726 2010-05-27 15:00:00   26     PM   240 121.000
6860  2010052726 2010-05-27 15:00:00   26     PM   245 121.000
6861  2010052726 2010-05-27 15:00:00   26     PM   250 121.000
6862  2010052726 2010-05-27 15:00:00   26     PM   255 121.000
6863  2010052726 2010-05-27 15:00:00   26     PM   260 121.000
6864  2010052726 2010-05-27 15:00:00   26     PM   265 121.000
6865  2010052726 2010-05-27 15:00:00   26     PM   270 121.000
6866  2010052726 2010-05-27 15:00:00   26     PM   275 121.000
6867  2010052726 2010-05-27 15:00:00   26     PM   280 121.000


 2018061364


id   DATETIME site period  lgth  wt_lbs
10841  2018061364 2018-06-13   64     PM   215  92.000
10842  2018061364 2018-06-13   64     PM   220  92.000
10843  2018061364 2018-06-13   64     PM   225  92.000
10844  2018061364 2018-06-13   64     PM   230  92.000
10845  2018061364 2018-06-13   64     PM   235  92.000
10846  2018061364 2018-06-13   64     PM   240  92.000
10847  2018061364 2018-06-13   64     PM   245  92.000
10848  2018061364 2018-06-13   64     PM   250  92.000
10849  2018061364 2018-06-13   64     PM   255  92.000
10850  2018061364 2018-06-13   64     PM   260  92.000
10851  2018061364 2018-06-13   64     PM   265  92.000
10852  2018061364 2018-06-13   64     PM   270  92.000

In [24]:
# it looks like 4018061364 is our only potential error (fixed earlier, now 2018061364)
df_LF.groupby('id').count().sort_values('yy')['yy'].head()

id
2007060235    1
1991051212    3
1994050848    4
2005050501    6
1991051312    6
Name: yy, dtype: int64

In [25]:
# loop through small counts above and look at them
# these look fine, unless there were multiple combined typos
smallgroups = list(df_LF.groupby('id').count().sort_values('yy')['yy'].head().index)

for group in smallgroups:
    idx = df_LF[df_LF.id == group].index.values[0]
    before = 2
    after = 8
    
    print('\n', df_LF[df_LF.id == group]['id'].values[0])
    display(df_LF.loc[idx-before:idx+after][['DATETIME', 'site', 'period', 'lgth', 'wt_lbs']])



 2007060235


DATETIME site period  lgth  wt_lbs
5960 2007-06-01   47     PM   285 133.000
5961 2007-06-01   47     PM   295 133.000
5962 2007-06-02   35     PM   225 122.000
5963 2007-06-02    2     PM   230 122.000
5964 2007-06-02    2     PM   235 122.000
5965 2007-06-02    2     PM   240 122.000
5966 2007-06-02    2     PM   245 122.000
5967 2007-06-02    2     PM   250 122.000
5968 2007-06-02    2     PM   255 122.000
5969 2007-06-02    2     PM   260 122.000
5970 2007-06-02    2     PM   265 122.000


 1991051212


DATETIME site period  lgth  wt_lbs
946 1990-06-08 00:00:00   64     AM   285     NaN
947 1990-06-08 00:00:00   64     AM   293     NaN
948 1991-05-12 03:00:00   12     AM   265   1.000
949 1991-05-12 03:00:00   12     AM   270   1.000
950 1991-05-12 03:00:00   12     AM   275   1.000
951 1991-05-13 03:00:00   12     AM   265  11.000
952 1991-05-13 03:00:00   12     AM   270  11.000
953 1991-05-13 03:00:00   12     AM   275  11.000
954 1991-05-13 03:00:00   12     AM   280  11.000
955 1991-05-13 03:00:00   12     AM   285  11.000
956 1991-05-13 03:00:00   12     AM   290  11.000


 1994050848


DATETIME site period  lgth  wt_lbs
10341 1994-06-01   41     AM   280 137.000
10342 1994-06-01   41     AM   300 137.000
10343 1994-05-08   48     AM   250   3.000
10344 1994-05-08   48     AM   255   3.000
10345 1994-05-08   48     AM   260   3.000
10346 1994-05-08   48     AM   265   3.000
10347 1994-06-02   49     PM   235 118.000
10348 1994-06-02   49     PM   240 118.000
10349 1994-06-02   49     PM   245 118.000
10350 1994-06-02   49     PM   250 118.000
10351 1994-06-02   49     PM   255 118.000


 2005050501


DATETIME site period  lgth  wt_lbs
5570 2004-06-16    5     PM   260  41.000
5571 2004-06-16    5     PM   265  41.000
5572 2005-05-05    1     PM   240   9.000
5573 2005-05-05    1     PM   250   9.000
5574 2005-05-05    1     PM   255   9.000
5575 2005-05-05    1     PM   260   9.000
5576 2005-05-05    1     PM   265   9.000
5577 2005-05-05    1     PM   270   9.000
5578 2005-05-07   26     PM   240  12.000
5579 2005-05-07   26     PM   245  12.000
5580 2005-05-07   26     PM   250  12.000


 1991051312


DATETIME site period  lgth  wt_lbs
949 1991-05-12 03:00:00   12     AM   270   1.000
950 1991-05-12 03:00:00   12     AM   275   1.000
951 1991-05-13 03:00:00   12     AM   265  11.000
952 1991-05-13 03:00:00   12     AM   270  11.000
953 1991-05-13 03:00:00   12     AM   275  11.000
954 1991-05-13 03:00:00   12     AM   280  11.000
955 1991-05-13 03:00:00   12     AM   285  11.000
956 1991-05-13 03:00:00   12     AM   290  11.000
957 1991-05-16 03:00:00   12     PM   255  11.000
958 1991-05-16 03:00:00   12     PM   260  11.000
959 1991-05-16 03:00:00   12     PM   265  11.000